### PR TITLE
[Dynamic Min Liq Filter] Implement and test liquidity pricer, worker and liquidity caching in tokens handler

### PR DESCRIPTION
**Pull request by p0mvn**
This PR focuses on implementing a pool liqudiity pricing worker that takes a map of toke liquidity and computes its market capitalization.

It also implements helpers on the tokens use case for storing and retrieving token liquidity capitalization.

This PR is separated from https://github.com/osmosis-labs/sqs/pull/224 for ease of review.

See feature documentation here: https://github.com/osmosis-labs/sqs/pull/281

### CODE REVIEW

File: domain/mvc/tokens.go [hunk 1]
```diff
@@ -7,8 +7,23 @@ import (
 	"github.com/osmosis-labs/sqs/domain"
 )
 
+type TokensPoolLiquidityHandler interface {
+	// GetChainScalingFactorByDenomMut returns a chain scaling factor for a given denom
+	// and a boolean flag indicating whether the scaling factor was found or not.
+	// Note that the returned decimal is a shared resource and must not be mutated.
+	// A clone should be made for any mutative operation.
+	GetChainScalingFactorByDenomMut(denom string) (osmomath.Dec, error)
```
Comment (cryptomatictrader) [inline domain/mvc/tokens.go:15]: Just question - what is scaling factor? Something related to exponent?

File: domain/pricing.go [hunk 2]
```diff
@@ -148,9 +133,79 @@ type PricingUpdateListener interface {
...
+	// RepriceDenomMetadata reprices the token liquidity metadata for the denoms updated within the block.
+	// Returns the updated token metadata.
+	// If there is an update for a denom with a later height than the current height, it is skipped, making this a no-op.
+	// Relies on the blockPriceUpdates to get the price for the denoms.
+	// If the price for denom cannot be fetched, the liquidity capitalization for this denom is set to zero.
+	// The latest update height for this denom is updated on completion.
+	RepriceDenomMetadata(updateHeight uint64, blockPriceUpdates PricesResult, quoteDenom string, blockDenomLiquidityUpdatesMap DenomPoolLiquidityMap) PoolDenomMetaDataMap
+
+	// GetHeightForDenom returns zero if the height is not found or fails to cast it to the return type.
+	GetHeightForDenom(denom string) uint64
```
Comment (cryptomatictrader) [inline domain/pricing.go]: Questions:
1. Does it mean we're keeping track of all the block heights which triggered updates for every denom?

File: tokens/usecase/pricing/worker/pool_liquidity_pricer_worker.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+	updateListeners []domain.PoolLiquidityComputeListener
+
+	liquidityPricer domain.LiquidityPricer
+
+	// Denom -> Last height of the pricing update.
+	// This exists because pricing computations are asynchronous. As a result, a pricing update for a later
+	// height might arrive before a pricing update for an earlier height. This map is used to ensure that
+	// the latest height pricing update for a denom is used.
+	latestHeightForDenom sync.Map
```
Comment (cryptomatictrader) [inline tokens/usecase/pricing/worker/pool_liquidity_pricer_worker.go:30]: 👍

File: domain/mvc/tokens.go (hunk 1 above)
Comment (p0mvn) [reply domain/mvc/tokens.go:15]: Yes, if exponent in the [asset list](https://github.com/osmosis-labs/assetlists/blob/main/osmosis-1/generated/frontend/assetlist.json) is 6, the scaling factor is 10^6

File: domain/pricing.go (hunk 2 above)
Comment (p0mvn) [reply domain/pricing.go]: No, we keep track of the latest height per denom.

I will rename to `GetLatestUpdateHeightForDenom` to reflect this

Comment (p0mvn) [reply domain/pricing.go]: [03565aa](https://github.com/osmosis-labs/sqs/pull/310/commits/03565aa62b9a520565591071b114e0f7b361009e)

**Comment by p0mvn** [comment]
@cryptomatictrader all current comments are now addressed

File: tokens/usecase/tokens_usecase.go [hunk 4]
```diff
@@ -113,11 +117,76 @@ func NewTokensUsecase(tokenMetadataByChainDenom map[string]domain.Token) mvc.Tok
...
+	poolDenomMetadataObj, ok := t.poolDenomMetaData.Load(chainDenom)
+	if !ok {
+		return domain.PoolDenomMetaData{}, domain.PoolDenomMetaDataNotPresentError{
+			ChainDenom: chainDenom,
+		}
+	}
+
+	poolDenomMetadata, ok := poolDenomMetadataObj.(domain.PoolDenomMetaData)
+	if !ok {
+		// TODO: add alert
```
Comment (p0mvn) [inline tokens/usecase/tokens_usecase.go]:
```suggestion
```

File: tokens/usecase/pricing/worker/pool_liquidity_pricer_worker.go [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+	p.tokenPoolLiquidityHandler.UpdatePoolDenomMetadata(repricedTokenMetadata)
+
+	// Notify listeners.
+	for _, listener := range p.updateListeners {
+		// Avoid checking error since we want to execute all listeners.
+		_ = listener.OnPoolLiquidityCompute(int64(height))
+	}
+
+	// Measure duration
+	domain.SQSPoolLiquidityPricingWorkerComputeDurationGauge.Add(float64(time.Since(start).Milliseconds()))
```
Comment (cryptomatictrader) [inline tokens/usecase/pricing/worker/pool_liquidity_pricer_worker.go:63]: This is good 👍

File: tokens/usecase/pricing/worker/pool_liquidity_pricer_worker.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+	// Note: in the future, if we add pool liquidity pricing, we can process the computation in separate goroutines
+	// for concurrency.
+	repricedTokenMetadata := p.RepriceDenomMetadata(height, baseDenomPriceUpdates, quoteDenom, blockPoolMetadata.DenomPoolLiquidityMap)
+
+	// Update the pool denom metadata.
+	p.tokenPoolLiquidityHandler.UpdatePoolDenomMetadata(repricedTokenMetadata)
+
+	// Notify listeners.
+	for _, listener := range p.updateListeners {
```
Comment (cryptomatictrader) [inline tokens/usecase/pricing/worker/pool_liquidity_pricer_worker.go:57]: Use separate thread for each listener in future? I think it's good for now as we don't have many listeners and using thread may just be an overkill.

File: tokens/usecase/tokens_usecase.go
Comment (cryptomatictrader) [inline tokens/usecase/tokens_usecase.go:1]: Thanks for the documentation - I understand what the code does in no time 👍

**Review by cryptomatictrader — APPROVED** [review]
Thanks for breaking it down for easier review. Nice work and it's good to go!

File: tokens/usecase/pricing/worker/pool_liquidity_pricer_worker.go (hunk 6 above)
Comment (p0mvn) [reply tokens/usecase/pricing/worker/pool_liquidity_pricer_worker.go:57]: Yes, I don't think that concurrency is necessary at this stage. The worker flow is already running in a separate goroutine.

One of the listeners is just a healthcheck update. Adding concurrency now would add more complexity without clear benefit.

In the future, we can revisit, if the scope of these workers changes

**Pull request merged**